// lac-mat-tel-input/esm2015/lib/lac-mat-tel-input.component.js
var LacMatTelInputComponent_1;
import { __decorate, __metadata, __param } from "tslib";
import { Component, HostBinding, Input, Optional, Self, ElementRef, ViewChild, EventEmitter, Output, ChangeDetectorRef } from '@angular/core';
import { MatFormFieldControl } from '@angular/material/form-field';
import { MatInput } from '@angular/material/input';
import { Subject } from 'rxjs';
import { NgControl } from '@angular/forms';
import { FocusMonitor } from '@angular/cdk/a11y';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { getExampleNumber, parsePhoneNumberFromString, AsYouType, getCountryCallingCode } from 'libphonenumber-js';
import { Examples } from './data/country-code';
let LacMatTelInputComponent = LacMatTelInputComponent_1 = class LacMatTelInputComponent {
    constructor(ngControl, fm, elRef, changeDetector) {
        this.ngControl = ngControl;
        this.fm = fm;
        this.elRef = elRef;
        this.changeDetector = changeDetector;
        this.blur = new EventEmitter();
        this.PhoneNumberMaxDigits = 20;
        this.maxInputLength = this.PhoneNumberMaxDigits;
        //Subject to notify when country needs to be update from input
        //Se paste condition in onInputChanged
        this.countryChange = new Subject();
        //Mat Form Field implementation - BEGIN
        this.stateChanges = new Subject();
        this.id = `lac-mat-tel-input-${LacMatTelInputComponent_1.nextId++}`;
        //ngControl: NgControl; set with Dependency Injection
        this.focused = false;
        //adds a required indicator to the label/placeholder
        this._required = false;
        this.inputDisabled = false;
        this.controlType = 'mat-tel';
        this.describedBy = '';
        this.onTouched = () => {
        };
        this.propagateChange = (_) => {
        };
        //ControlValueAccessor - END
        this._internationalFormat = false;
        if (this.ngControl != null) {
            // Setting the value accessor directly (instead of using
            // the providers) to avoid running into a circular import.
            this.ngControl.valueAccessor = this;
        }
        fm.monitor(elRef.nativeElement, true).subscribe(origin => {
            this.focused = !!origin;
            this.stateChanges.next();
        });
    }
    set value(v) {
        this._value = v ? `+${getCountryCallingCode(this.selectedCountry)} ${v}` : v;
        if (this.internationalFormat && this._value) {
            let phoneNumber = parsePhoneNumberFromString(this._value);
            if (phoneNumber && phoneNumber.isValid()) {
                this._value = phoneNumber.formatInternational();
            }
        }
        this.propagateChange(this._value);
        this.stateChanges.next();
    }
    ;
    get placeholder() {
        return this._placeholder;
    }
    ;
    set placeholder(val) {
        this._placeholder = val;
        this.stateChanges.next();
    }
    get empty() {
        return !this.phone;
    }
    ;
    //shouldLabelFloat: boolean;
    get shouldLabelFloat() {
        return this.focused || !this.empty;
    }
    get required() {
        return this._required;
    }
    set required(req) {
        this._required = coerceBooleanProperty(req);
        this.stateChanges.next();
    }
    get disabled() { return this.inputDisabled || this.ngControl.disabled; }
    set disabled(value) {
        this.inputDisabled = coerceBooleanProperty(value);
        this.stateChanges.next();
    }
    get errorState() {
        return this.ngControl && this.ngControl.errors !== null && !!this.ngControl.touched;
    }
    setDescribedByIds(ids) {
        this.describedBy = ids.join(' ');
    }
    onContainerClick(event) {
        if (!this.disabled) {
            this.onTouched();
        }
    }
    //Mat Form Field implementation - END
    //ControlValueAccessor - BEGIN
    writeValue(value) {
        this.onInputChanged(value);
    }
    registerOnChange(fn) {
        this.propagateChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    get internationalFormat() {
        return this._internationalFormat;
    }
    set internationalFormat(format) {
        this._internationalFormat = coerceBooleanProperty(format);
        this.onInputChanged(this.phone);
    }
    //lifecycle
    ngOnInit() {
    }
    ngOnDestroy() {
        this.fm.stopMonitoring(this.elRef.nativeElement);
        this.countryChange.complete();
        this.stateChanges.complete();
    }
    onCountrySelected(code, noPhoneReset) {
        let hasCountryChanged = this.selectedCountry && this.selectedCountry != code;
        this.selectedCountry = code;
        let example = getExampleNumber(code, Examples).formatNational();
        let numbersOnly = example.replace(/[^\d]/g, '');
        let maxExample = numbersOnly + '9999999999';
        this.maxInputLength = this.PhoneNumberMaxDigits;
        //find out the maximum (formatted) size a valid number for the country can have
        for (let i = maxExample.length; i >= 0; i--) {
            let test = maxExample.substring(0, i);
            let testPhone = parsePhoneNumberFromString(test, code);
            if (testPhone && testPhone.isValid()) {
                let maxInput = new AsYouType(code).input(test);
                this.maxInputLength = maxInput.length;
                this.placeholder = maxInput; //TODO set placeholder optionally
                break;
            }
        }
        //when new country is selected reset phone if phone reset was not disabled 
        if (hasCountryChanged && !noPhoneReset) {
            this.phone = '';
            this.value = this.phone;
            //focus on input
            if (this.phoneInput) {
                setTimeout(() => this.phoneInput.focus());
            }
        }
    }
    onInputChanged(e) {
        if (!e) {
            e = '';
        }
        this.phone = e;
        if (e.startsWith('+')) { //handles pasting of a complete international number
            try {
                let pastedNumber = parsePhoneNumberFromString(e);
                if (pastedNumber && pastedNumber.country) {
                    let code = pastedNumber.country;
                    this.onCountrySelected(code, true);
                    this.countryChange.next(code);
                    setTimeout(() => {
                        this.phone = pastedNumber.formatNational();
                        this.value = this.phone;
                    });
                    return;
                }
            }
            catch (_a) {
            }
        }
        let numbersOnly = e.replace(/[^\d]/g, '');
        setTimeout(() => {
            let formatted = new AsYouType(this.selectedCountry).input(numbersOnly);
            //if the formatted output equals what we already have then the user is trying to delete a symbol inserted by
            //the formatted version
            this.phone = formatted.substr(0, formatted.length - 1) === this.phone ? numbersOnly : formatted;
            this.value = this.phone;
        }, 0);
    }
    onInputKeyPress(event) {
        const allowedChars = /[0-9\+\-\ ]/;
        const allowedCtrlChars = /[axcv]/; // Allows copy-pasting
        const allowedOtherKeys = [
            'ArrowLeft', 'ArrowUp', 'ArrowRight', 'ArrowDown',
            'Home', 'End', 'Insert', 'Delete', 'Backspace'
        ];
        if (!allowedChars.test(event.key)
            && !(event.ctrlKey && allowedCtrlChars.test(event.key))
            && !(allowedOtherKeys.includes(event.key))) {
            event.preventDefault();
        }
    }
};
//the id of an element to associate labels and hints with.
LacMatTelInputComponent.nextId = 0;
LacMatTelInputComponent.ctorParameters = () => [
    { type: NgControl, decorators: [{ type: Optional }, { type: Self }] },
    { type: FocusMonitor },
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], LacMatTelInputComponent.prototype, "blur", void 0);
__decorate([
    ViewChild('phoneInput', { static: true }),
    __metadata("design:type", MatInput)
], LacMatTelInputComponent.prototype, "phoneInput", void 0);
__decorate([
    HostBinding(),
    __metadata("design:type", String)
], LacMatTelInputComponent.prototype, "id", void 0);
__decorate([
    HostBinding('class.floating'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], LacMatTelInputComponent.prototype, "shouldLabelFloat", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], LacMatTelInputComponent.prototype, "required", null);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], LacMatTelInputComponent.prototype, "disabled", null);
__decorate([
    HostBinding('attr.aria-describedby'),
    __metadata("design:type", Object)
], LacMatTelInputComponent.prototype, "describedBy", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], LacMatTelInputComponent.prototype, "internationalFormat", null);
LacMatTelInputComponent = LacMatTelInputComponent_1 = __decorate([
    Component({
        selector: 'lac-mat-tel-input',
        template: "<input type=\"text\" matInput #phoneInput=\"matInput\"\n  [ngModel]=\"phone\"\n  (blur)=\"blur.emit($event)\"\n  (keypress)=\"onInputKeyPress($event)\"\n  (ngModelChange)=\"onInputChanged($event)\"\n  [disabled]=\"disabled\"\n  [maxlength]=\"!phone ? PhoneNumberMaxDigits : maxInputLength\"\n  [placeholder]=\"placeholder\"  />\n",
        providers: [{ provide: MatFormFieldControl, useExisting: LacMatTelInputComponent_1 }],
        styles: [""]
    }),
    __param(0, Optional()), __param(0, Self()),
    __metadata("design:paramtypes", [NgControl,
        FocusMonitor,
        ElementRef,
        ChangeDetectorRef])
], LacMatTelInputComponent);
export { LacMatTelInputComponent };
